### ISSUE TITLE
ping: EINTR on aarch64 causes exit 1

### ISSUE THREAD

**Issue by pevik** [iputils/iputils#424]
Originally put in https://github.com/iputils/iputils/issues/423#issuecomment-1249197841, but it looks to be a separate issue.

LTP test run on aarch64 found a random failures in `int close_stream(FILE *stream)`:
https://openqa.opensuse.org/tests/2664916#step/if4-mtu-change_ifconfig/8

/opt/ltp/testcases/bin/tst_test.sh: line 156: 25324 Aborted                 (core dumped) tst_rod "$@"
if-mtu-change 1 TFAIL: ping -I ltp_ns_veth2 -c 500 -s 750 -f 10.0.0.1 >/dev/null failed unexpectedly

For a record sources of `tst_ping`
https://github.com/linux-test-project/ltp/blob/98140f3d6a374791c9dd4d38425a239d02a47b29/testcases/lib/tst_net.sh#L853-L902

```
# strace -o /tmp/ping.$$ -f ping  -I ltp_ns_veth2 -c 500 -s 392 -f 10.0.0.1
# OK
...
setitimer(ITIMER_REAL, {it_interval={tv_sec=0, tv_usec=0}, it_value={tv_sec=0, tv_usec=618}}, NULL) = 0
recvmsg(3, {msg_name={sa_family=AF_INET, sin_port=htons(0), sin_addr=inet_addr("10.0.0.1")}, msg_namelen=128 => 16, msg_iov=[{iov_base="\0\0-\311\6_\1\364n1#c\0\0\0\0k\373\1\0\0\0\0\0\20\21\22\23\24\25\26\27"..., iov_len=65643}], msg_iovlen=1, msg_control=[{cmsg_len=32, cmsg_level=SOL_SOCKET, cmsg_type=SO_TIMESTAMP_OLD, cmsg_data={tv_sec=1663250798, tv_usec=130145}}, {cmsg_len=20, cmsg_level=SOL_IP, cmsg_type=IP_TTL, cmsg_data=[64]}], msg_controllen=56, msg_flags=0}, 0) = 65515
write(1, "\10 \10", 3)                  = 3
write(1, "\n", 1)                       = 1
write(1, "--- 10.0.0.1 ping statistics ---"..., 33) = 33
write(1, "500 packets transmitted, ", 25) = 25
write(1, "500 received", 12)            = 12
write(1, ", 0% packet loss", 16)        = 16
write(1, ", time 319ms", 12)            = 12
--- SIGALRM {si_signo=SIGALRM, si_code=SI_KERNEL} ---
rt_sigreturn({mask=[]})                 = 12
write(1, "\n", 1)                       = 1
write(1, "rtt min/avg/max/mdev = 0.246/0.2"..., 49) = 49
write(1, ", ipg/ewma 0.639/0.251 ms", 25) = 25
write(1, "\n", 1)                       = 1
brk(0xaaaad780e000)                     = 0xaaaad780e000
close(1)                                = 0
close(2)                                = 0
exit_group(0)                           = ?
+++ exited with 0 +++

# ERR
...
setitimer(ITIMER_REAL, {it_interval={tv_sec=0, tv_usec=0}, it_value={tv_sec=0, tv_usec=244}}, NULL) = 0
recvmsg(3, {msg_name={sa_family=AF_INET, sin_port=htons(0), sin_addr=inet_addr("10.0.0.1")}, msg_namelen=128 => 16, msg_iov=[{iov_base="\0\08\205\31\4\1\364\242I$c\0\0\0\0t\264\v\0\0\0\0\0\20\21\22\23\24\25\26\27"..., iov_len=528}], msg_iovlen=1, msg_control=[{cmsg_len=32, cmsg_level=SOL_SOCKET, cmsg_type=SO_TIMESTAMP_OLD, cmsg_data={tv_sec=1663322530, tv_usec=767174}}, {cmsg_len=20, cmsg_level=SOL_IP, cmsg_type=IP_TTL, cmsg_data=[64]}], msg_controllen=56, msg_flags=0}, 0) = 400
write(1, "\10 \10", 3)            = 3
write(1, "\n", 1)                 = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
--- SIGALRM {si_signo=SIGALRM, si_code=SI_KERNEL} ---
rt_sigreturn({mask=[]})           = -1 EINTR (Interrupted system call)
write(1, "--- 10.0.0.1 ping statistics ---"..., 33) = 33
write(1, "500 packets transmitted, ", 25) = 25
write(1, "500 received", 12)      = 12
write(1, ", 0% packet loss", 16)  = 16
write(1, ", time 223ms", 12)      = 12
write(1, "\n", 1)                 = 1
write(1, "rtt min/avg/max/mdev = 0.070/0.0"..., 49) = 49
write(1, ", ipg/ewma 0.446/0.083 ms", 25) = 25
write(1, "\n", 1)                 = 1
close(1)                          = 0
write(2, "ping: ", 6)             = 6
write(2, "write error", 11)       = 11
write(2, "\n", 1)                 = 1
exit_group(1)                     = ?
+++ exited with 1 +++
```

Sometimes there is `ERESTARTSYS` followed by `SIGALRM` and `EINTR`, which ends up exiting `1` due `write error` when closing `stdout` (no errno specified):

https://github.com/iputils/iputils/blob/1ab5fafee6261369d962eb27e2fe08c8418bd74e/iputils_common.c#L58-L69

Debugging shows that `EINTR` is causing error in `close_stream()`, I wonder if it should be also whitelisted (currently `EPIPE` and `EBADF` are whitelisted):

https://github.com/iputils/iputils/blob/1ab5fafee6261369d962eb27e2fe08c8418bd74e/iputils_common.c#L36-L56

I'm not sure if adding `close_stream()` was a good idea at all (4655ecc5105c383669ef529f21f3344f99e7372f).

**Comment by gobenji** [comment]
From experimentation, it looks like write() -> EINTR results in ferror() being set and remaining set. That explains why close_stream() returns an error in the "# ERR" case (prev_fail is set).

How about setting SA_RESTART when doing sigaction(SIGALRM, ...). It avoids EINTR for write(). Maybe it would also be appropriate for the other signals configured in setup().

**Comment by gobenji** [comment]
> How about setting SA_RESTART when doing sigaction(SIGALRM, ...). It avoids EINTR for write(). Maybe it would also be appropriate for the other signals configured in setup().

On second thought I'm afraid that would prevent the quick termination by a signal of a `ping` that is blocked on some IO. So setting SA_RESTART is not an adequate solution.

**Comment by pevik** [comment]
Yes, you're right `.sa_flags = SA_RESTART` could blocks. I wonder if just whitelisting `EINTR` would be better.

**Comment by metan-ucw** [comment]
If you look at the ERR trace closely you can see that piece of the output is actually missing, hence the non-zero exit status is actually correct. When this issue happens random part of the output is missing, which is unexpectedly hard to spot.

So I would say that we either have to live with `SA_RESTART` or incomplete output, unless we manage to rewrite the code in a saner way.

**Comment by pevik** [comment]
I was blind, as @metan-ucw pointed out there is really missing output, thus whitelisting `EINTR` is obviously wrong idea.

**Comment by pevik** [comment]
@gobenji Actually `SA_RESTART` works in ping code, I was able to reproduce segfault before, but now it does not segfault and don't block.
